### PR TITLE
Equivocation misbehaviour reports

### PR DESCRIPTION
**Pull request by marcio-diaz**
Replaces #2800, fixing conflicts, adding tests and improving code.

TODO: Test that equivocations really do lead to reports in GRANDPA (no idea how to check it in a meaningful way since equivocations come from grandpa repo), Babe stuff (maybe follow up).

### CODE REVIEW

File: core/test-runtime/src/lib.rs [hunk 1]
```diff
@@ -459,9 +467,14 @@ cfg_if! {
...
 			}
 
-			impl consensus_aura::AuraApi<Block, AuraId> for Runtime {
+			impl consensus_aura::AuraApi<Block, AuraId, sr25519::Signature> for Runtime {
 				fn slot_duration() -> u64 { 1 }
 				fn authorities() -> Vec<AuraId> { system::authorities() }
+				fn construct_equivocation_report_call(
+					_proof: AuraEquivocationProof<<Block as BlockT>::Header,sr25519::Signature>
+				) -> Vec<u8> {
+					vec![]
```
Comment (DemiMarie-parity) [inline core/test-runtime/src/lib.rs]: I was wondering why this was empty, then realized that this is the _test_ runtime.

File: node-template/runtime/src/lib.rs [hunk 2]
```diff
@@ -261,13 +263,19 @@ impl_runtime_apis! {
...
 			Aura::slot_duration()
 		}
 		fn authorities() -> Vec<AuraId> {
 			Aura::authorities()
 		}
+
+		fn construct_equivocation_report_call(
+			_proof: AuraEquivocationProof<<Block as BlockT>::Header,ed25519::Signature>
+		) -> Vec<u8> {
+			vec![]
```
Comment (DemiMarie-parity) [inline node-template/runtime/src/lib.rs]: This, however, should not be empty.  It seems to me that a proof is definitely required.  Even if it is not, there should be documentation explaining why the empty vector is a valid choice.

File: srml/aura/src/lib.rs [hunk 3]
```diff
@@ -167,7 +242,15 @@ decl_storage! {
...
 
 decl_module! {
-	pub struct Module<T: Trait> for enum Call where origin: T::Origin { }
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin { 
+		/// Report equivocation.
+		fn report_equivocation(
+			_origin,
+			_equivocation_proof: AuraEquivocationProof<T::Header, T::Signature>
+		) {
+			// This is the place where we will slash.
```
Comment (DemiMarie-parity) [inline srml/aura/src/lib.rs]: This needs an issue filed for it, along with a FIXME and the issue number in the code.

File: srml/aura/src/tests.rs [hunk 4]
```diff
@@ -90,3 +98,64 @@ fn aura_reports_offline() {
...
+
+	let proof3 = AuraEquivocationProof::new(header1.clone(), header2.clone(), sig1.clone(), sig1.clone());
+	let call3 = Call::report_equivocation(proof3);
+
+	with_externalities(&mut new_test_ext(authorities), || {
+		assert_eq!(Aura::validate_unsigned(&call1), TransactionValidity::Valid {
+			priority: 10,
+			requires: vec![],
+			provides: vec![],
+			longevity: 18446744073709551615,
```
Comment (DemiMarie-parity) [inline srml/aura/src/tests.rs]:
```suggestion
			longevity: std::u64::MAX,
```

File: srml/babe/src/lib.rs [hunk 5]
```diff
@@ -117,7 +129,26 @@ decl_storage! {
...
 
 decl_module! {
-	pub struct Module<T: Trait> for enum Call where origin: T::Origin { }
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Report equivocation.
+		fn report_equivocation(
+			_origin,
+			_equivocation_proof: BabeEquivocationProof<T::Header, T::Signature>
+		) {
+			// This is the place where we will slash.
```
Comment (DemiMarie-parity) [inline srml/babe/src/lib.rs]: Again, this needs a FIXME and an issue number.

File: srml/grandpa/src/lib.rs [hunk 6]
```diff
@@ -151,10 +159,20 @@ decl_module! {
...
-		/// Report some misbehavior.
-		fn report_misbehavior(origin, _report: Vec<u8>) {
-			ensure_signed(origin)?;
-			// FIXME: https://github.com/paritytech/substrate/issues/1112
+		/// Report prevote equivocation in Grandpa.
+		fn report_prevote_equivocation(
+			_origin,
+			_equivocation_proof: GrandpaEquivocationProof<PrevoteEquivocation<T::Hash, T::BlockNumber>>
+		) {
+			// Slash
```
Comment (DemiMarie-parity) [inline srml/grandpa/src/lib.rs]: Ditto.

File: srml/grandpa/src/lib.rs [hunk 7]
```diff
@@ -151,10 +159,20 @@ decl_module! {
...
+		}
+
+		/// Report precommit equivocation in Grandpa.
+		fn report_precommit_equivocation(
+			_origin,
+			_equivocation_proof: GrandpaEquivocationProof<PrecommitEquivocation<T::Hash, T::BlockNumber>>
+		) {
+			// Slash
```
Comment (DemiMarie-parity) [inline srml/grandpa/src/lib.rs]: Here too

File: srml/grandpa/src/lib.rs [hunk 8]
```diff
@@ -299,3 +317,112 @@ impl<T: Trait> finality_tracker::OnFinalizationStalled<T::BlockNumber> for Modul
...
+		second_payload.as_slice(),
+		&equivocation.identity
+	);
+
+	if valid_signature1 && valid_signature2 && first_message != second_message {
+		return TransactionValidity::Valid {
+			priority: 10,
+			requires: vec![],
+			provides: vec![],
+			longevity: 18446744073709551615,
```
Comment (DemiMarie-parity) [inline srml/grandpa/src/lib.rs]:
```suggestion
			longevity: u64::MAX,
```

File: srml/grandpa/src/lib.rs [hunk 9]
```diff
@@ -299,3 +317,112 @@ impl<T: Trait> finality_tracker::OnFinalizationStalled<T::BlockNumber> for Modul
...
```
Comment (DemiMarie-parity) [inline srml/grandpa/src/lib.rs]:
```suggestion
			longevity: u64::MAX,
```

File: srml/grandpa/src/tests.rs [hunk 10]
```diff
@@ -202,3 +203,141 @@ fn dispatch_forced_change() {
...
+	let call1 = Call::report_prevote_equivocation(proof1);
+	let call2 = Call::report_prevote_equivocation(proof2);
+	let call3 = Call::report_prevote_equivocation(proof3);
+
+	with_externalities(&mut new_test_ext_sig(authorities), || {
+		assert_eq!(Grandpa::validate_unsigned(&call1), TransactionValidity::Valid {
+			priority: 10,
+			requires: vec![],
+			provides: vec![],
+			longevity: 18446744073709551615,
```
Comment (DemiMarie-parity) [inline srml/grandpa/src/tests.rs]:
```suggestion
			longevity: u64::MAX,
```

File: srml/grandpa/src/tests.rs [hunk 11]
```diff
@@ -202,3 +203,141 @@ fn dispatch_forced_change() {
...
+	let call1 = Call::report_precommit_equivocation(proof1);
+	let call2 = Call::report_precommit_equivocation(proof2);
+	let call3 = Call::report_precommit_equivocation(proof3);
+
+	with_externalities(&mut new_test_ext_sig(authorities), || {
+		assert_eq!(Grandpa::validate_unsigned(&call1), TransactionValidity::Valid {
+			priority: 10,
+			requires: vec![],
+			provides: vec![],
+			longevity: 18446744073709551615,
```
Comment (DemiMarie-parity) [inline srml/grandpa/src/tests.rs]:
```suggestion
			longevity: u64::MAX,
```

File: srml/aura/src/lib.rs [hunk 12]
```diff
@@ -154,6 +163,72 @@ pub trait Trait: timestamp::Trait {
...
+
+		if !proof.second_signature().verify(pre_hash2.as_ref(), author) {
+			return TransactionValidity::Invalid(0)
+		}
+
+		return TransactionValidity::Valid {
+			priority: 10,
+			requires: vec![],
+			provides: vec![],
+			longevity: 18446744073709551615,
```
Comment (DemiMarie-parity) [inline srml/aura/src/lib.rs]:
```suggestion
			longevity: u64::MAX,
```

File: core/finality-grandpa/primitives/src/lib.rs [hunk 13]
```diff
@@ -115,5 +126,25 @@ decl_runtime_apis! {
...
+		
+		/// Construct a call to report the precommit equivocation.
+		fn construct_precommit_equivocation_report_call(
+			proof: GrandpaEquivocationProof<PrecommitEquivocation<<Block as BlockT>::Hash, NumberFor<Block>>>
+		) -> Vec<u8>;
 	}
 }
+
+#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+pub struct GrandpaEquivocationProof<E> {
```
Comment (rphmeier) [inline core/finality-grandpa/primitives/src/lib.rs]: docs on public items (including their public fields)

File: core/finality-grandpa/primitives/src/lib.rs [hunk 14]
```diff
@@ -115,5 +126,25 @@ decl_runtime_apis! {
...
+	pub set_id: u64,
+	pub equivocation: E,
+}
+
+pub fn localized_payload<E: Encode>(round: u64, set_id: u64, message: &E) -> Vec<u8> {
```
Comment (rphmeier) [inline core/finality-grandpa/primitives/src/lib.rs]: docs on public items

File: core/consensus/aura/primitives/src/lib.rs [hunk 15]
```diff
@@ -45,5 +49,70 @@ decl_runtime_apis! {
...
+		fn construct_equivocation_report_call(
+			proof: AuraEquivocationProof<<Block as BlockT>::Header, Signature>,
+		) -> Vec<u8>;
+	}
+}
+
+/// Get slot author for given block along with authorities.
+pub fn slot_author<AuthorityId>(slot_num: u64, authorities: &[AuthorityId]) -> Option<&AuthorityId>
+{
+	if authorities.is_empty() { return None }
```
Comment (niklasad1) [inline core/consensus/aura/primitives/src/lib.rs]: I don't understand why you need these extra checks when the function returns an `Option` and `get` does it for you.

why not just 
```rustc
pub fn slot_author<AuthorityId>(slot_num: u64, authorities: &[AuthorityId]) -> Option<&AuthorityId> {
   authorities.get(slot_num as usize)
}
```

File: core/consensus/aura/primitives/src/lib.rs [hunk 16]
```diff
@@ -45,5 +49,70 @@ decl_runtime_apis! {
...
+
+	let idx = slot_num % (authorities.len() as u64);
```
Comment (niklasad1) [inline core/consensus/aura/primitives/src/lib.rs]: why `wrap-around` slot number here to fit in the `authorities slice`?

**Comment by andresilva** [comment]
Is this good to be reviewed? Could you merge/rebase master if that's the case? :)

File: core/consensus/aura/primitives/src/lib.rs [hunk 17]
```diff
@@ -45,5 +49,70 @@ decl_runtime_apis! {
...
+
+	let current_author = authorities.get(idx as usize)
+		.expect("authorities not empty; index constrained to list length;\
+				this is a valid index; qed");
+
+	Some(current_author)
+}
+
+#[derive(Debug, Encode, Decode, PartialEq, Eq, Clone)]
+pub struct AuraEquivocationProof<H, S> {
```
Comment (rphmeier) [inline core/consensus/aura/primitives/src/lib.rs]: docs on public items

File: srml/aura/src/lib.rs [hunk 18]
```diff
@@ -154,6 +163,72 @@ pub trait Trait: timestamp::Trait {
...
+	}
+
+	let maybe_slot1 = find_pre_digest::<_, T::Signature>(header1);
+	let maybe_slot2 = find_pre_digest::<_, T::Signature>(header2);
+
+	if maybe_slot1.is_ok() && maybe_slot1 == maybe_slot2 {
+		let slot = maybe_slot1.expect("OK by previous line; qed");
+
+		// FIX: What if there is a different set of authorities?
+		let authorities = <Module<T>>::authorities();
```
Comment (rphmeier) [inline srml/aura/src/lib.rs]: right, this is the hardest part of slashing. we probably will need the historical session trie to also store some key mapped to `Vec<T::Keys>`. Consensus engines that need access to the entire authorities list will have to accept a proof of that branch as well.

**Comment by DemiMarie-parity** [comment]
@marcio-diaz Can you rebase this?

**Comment by marcio-diaz** [comment]
Closing in favor of #3225 (between other follow ups)

**Pull request closed without merging**